pkg/agent/api/delegatedidentity/v1/service.go

```diff
@@ -0,0 +1,228 @@
+type Service struct {
+	delegatedidentityv1.UnsafeDelegatedIdentityServer
+
+	manager  manager.Manager
+	attestor Attestor
+
+	// Spiffe IDs of users that are authorized to use this API
+	authorizedUsers map[string]bool
+}
+
+// isCallerAuthorized attests the caller and returns true if its identity is on
+// the authorized users map.
+func (s *Service) isCallerAuthorized(ctx context.Context) (bool, error) {
+	callerSelectors, err := s.attestor.Attest(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	identities := s.manager.MatchingIdentities(callerSelectors)
+
+	for _, identity := range identities {
+		id := identity.Entry.SpiffeId
+
+		if _, ok := s.authorizedUsers[id]; ok {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (s *Service) SubscribeToX509SVIDs(req *delegatedidentityv1.SubscribeToX509SVIDsRequest, stream delegatedidentityv1.DelegatedIdentity_SubscribeToX509SVIDsServer) error {
+	ctx := stream.Context()
+	log := rpccontext.Logger(ctx)
+
+	authorized, err := s.isCallerAuthorized(ctx)
+
+	if err != nil {
+		log.WithError(err).Error("Workload attestation failed")
+		return status.Error(codes.Internal, "workload attestation failed")
+	}
+
+	if !authorized {
+		return status.Error(codes.PermissionDenied, "no authorized")
+	}
+
+	selectors, err := api.SelectorsFromProto(req.Selectors)
+	if err != nil {
+		return err
+	}
+
+	subscriber := s.manager.SubscribeToCacheChanges(selectors)
+	defer subscriber.Finish()
+
+	for {
+		select {
+		case update := <-subscriber.Updates():
+			if err := sendX509SVIDResponse(update, stream, log); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func sendX509SVIDResponse(update *cache.WorkloadUpdate, stream delegatedidentityv1.DelegatedIdentity_SubscribeToX509SVIDsServer, log logrus.FieldLogger) (err error) {
+	resp, err := composeX509SVIDBySelectors(update)
+	if err != nil {
+		log.WithError(err).Error("Could not serialize X.509 SVID response")
+		return status.Error(codes.Unavailable, "could not serialize response")
+	}
+
+	if err := stream.Send(resp); err != nil {
+		log.WithError(err).Error("Failed to send X.509 SVID response")
+		return err
+	}
+
+	return nil
+}
+
+func composeX509SVIDBySelectors(update *cache.WorkloadUpdate) (*delegatedidentityv1.SubscribeToX509SVIDsResponse, error) {
+	resp := new(delegatedidentityv1.SubscribeToX509SVIDsResponse)
+	resp.X509Svids = []*delegatedidentityv1.X509SVIDWithKey{}
+
+	for td, _ := range update.FederatedBundles {
+		resp.FederatesWith = append(resp.FederatesWith, td.IDString())
+	}
+
+	for _, identity := range update.Identities {
+		// TODO:  it doesn't work and always prints "false"
+		//fmt.Printf("processing id %s, %t\n", identity.Entry.SpiffeId, identity.Entry.Admin)
+		// Do not send admin nor downstream SVIDs to the caller
+		if identity.Entry.Admin || identity.Entry.Downstream {
+			continue
+		}
+
+		id, _ := idutil.IDProtoFromString(identity.Entry.SpiffeId)
```

Comment:
Even though we should really not ever hit an error here, we should still check for it because it could mean the difference between a bug affecting availability of an identity or a bug crashing the agent.

Reply:
Is this way OK?
```
id, err := idutil.IDProtoFromString(identity.Entry.SpiffeId)
if err != nil {
    return nil, fmt.Errorf("error during Spiffe ID parsing: %w", err)
}
```